crates/ruff_python_parser/src/string.rs

```diff
@@ -535,68 +227,95 @@ impl<'a> StringParser<'a> {
         }))
     }
 
-    fn parse(&mut self) -> Result<Vec<Expr>, LexicalError> {
-        if self.kind.is_any_fstring() {
-            self.parse_fstring(0)
-        } else if self.kind.is_any_bytes() {
-            self.parse_bytes().map(|expr| vec![expr])
+    fn parse(&mut self) -> Result<Expr, LexicalError> {
+        if self.kind.is_any_bytes() {
+            self.parse_bytes()
         } else {
-            self.parse_string().map(|expr| vec![expr])
+            self.parse_string()
         }
     }
 }
 
-fn parse_fstring_expr(source: &str, location: TextSize) -> Result<Expr, ParseError> {
-    let fstring_body = format!("({source})");
-    parse_expression_starts_at(&fstring_body, "<fstring>", location)
+pub(crate) fn parse_string_literal(
+    start: TextSize,
+    (source, kind, triple_quoted): (String, StringKind, bool),
+) -> Result<Expr, LexicalError> {
+    let start = start
+        + kind.prefix_len()
+        + if triple_quoted {
+            TextSize::from(3)
+        } else {
+            TextSize::from(1)
+        };
+    StringParser::new(source.as_str(), kind, start).parse()
 }
 
-fn parse_string(
-    source: &str,
-    kind: StringKind,
-    triple_quoted: bool,
+pub(crate) fn parse_fstring_middle(
     start: TextSize,
-) -> Result<Vec<Expr>, LexicalError> {
-    StringParser::new(source, kind, triple_quoted, start).parse()
+    (source, is_raw): (String, bool),
+) -> Result<Option<Expr>, LexicalError> {
+    // This is to account for the empty `FStringMiddle` token that is created
+    // to check for non-parenthesized lambda expressions.
+    if source.is_empty() {
```

Comment:
Could you expand this comment with an example and why it is necessary? I know we talked about it on Discord but new and external maintainers won't know about the conversation or don't have access to it.

Overall: this is unfortunate because it requires moving all parts into a new Vec (flatten, then collect). I wonder if a dedicated token would be a better solution